scripts/prepdocs/files.py

```diff
@@ -0,0 +1,749 @@
+        return len(encoding.encode(text))
+
+    def split_text_into_batches(self, texts: List[str]) -> List[List[str]]:
+        batch_info = OpenAIEmbeddings.SUPPORTED_BATCH_AOAI_MODEL.get(self.open_ai_model_name)
+        if not batch_info:
+            raise NotImplementedError(
+                f"Model {self.open_ai_model_name} is not supported with batch embedding operations"
+            )
+
+        batch_token_limit = batch_info["token_limit"]
+        batch_max_size = batch_info["max_batch_size"]
+        batches = []
+        batch = []
+        batch_token_length = 0
+        for text in texts:
+            text_token_length = self.calculate_token_length(text)
+            if batch_token_length + text_token_length >= batch_token_limit and len(batch) > 0:
+                batches.append(batch)
+                batch = []
+                batch_token_length = 0
+
+            batch.append(text)
+            batch_token_length = batch_token_length + text_token_length
+            if len(batch) == batch_max_size:
+                batches.append(batch)
+                batch = []
+                batch_token_length = 0
+
+        if len(batch) > 0:
+            batches.append(batch)
+
+        return batches
+
+    async def create_embedding_batch(self, texts: List[str]) -> List[List[float]]:
+        batches = self.split_text_into_batches(texts)
+        for batch in batches:
+            async for attempt in AsyncRetrying(
+                retry=retry_if_exception_type(openai.error.RateLimitError),
+                wait=wait_random_exponential(min=15, max=60),
+                stop=stop_after_attempt(15),
+                before_sleep=self.before_retry_sleep,
+            ):
+                with attempt:
+                    emb_args = await self.create_embedding_arguments()
+                    emb_response = await openai.Embedding.acreate(**emb_args, input=batch)
+                    return [data["embedding"] for data in emb_response["data"]]
+
+    async def create_embedding_single(self, text: str) -> List[float]:
+        async for attempt in AsyncRetrying(
+            retry=retry_if_exception_type(openai.error.RateLimitError),
+            wait=wait_random_exponential(min=15, max=60),
+            stop=stop_after_attempt(15),
+            before_sleep=self.before_retry_sleep,
+        ):
+            with attempt:
+                emb_args = await self.create_embedding_arguments()
+                emb_response = await openai.Embedding.acreate(**emb_args, input=text)
+                return emb_response["data"][0]["embedding"]
+
+    async def create_embeddings(self, texts: List[str]) -> List[List[float]]:
+        if not self.disable_batch and self.open_ai_model_name in OpenAIEmbeddings.SUPPORTED_BATCH_AOAI_MODEL:
+            return await self.create_embedding_batch(texts)
+
+        return [await self.create_embedding_single(text) for text in texts]
+
+
+class AzureOpenAIEmbeddingService(OpenAIEmbeddings):
```

Comment:
I wonder if we'll be able to replace these once OpenAI releases their latest SDK with OpenAi and AzureOpenAI. We'll see! Just a musing.

Reply:
yes, that will be a good follow up